lib/spack/spack/solver/concretize.lp

```diff
@@ -1077,6 +1088,30 @@ variant_is_propagated(PackageNode, Variant) :-
    attr("variant_selected", PackageNode, Variant, Value, _, _),
    not propagate(PackageNode, variant_value(Variant, Value)).
 
+% Cannot propagate the same variant from two distinct sources
+error(100, "{0} and {1} cannot both propagate variant '{2}'", Source1, Source2, Variant) :-
+  variant_is_propagated(node(X, Source1), variant_value(Variant, _)),
+  variant_is_propagated(node(Y, Source2), variant_value(Variant, _)),
+  Source1 < Source2.
+
+% Cannot propagate a variant if a different value was set for it in a dependency
+error(100, "Cannot propagate the variant '{0}' from the package: {1} because package: {2} is set to exclude it", Variant, Source, Package) :-
+  % Package has a variant set to Value1
+  attr("variant_set", node(X, Package), Variant, Value1),
+
+  % A different value is being propagated from somewhere else
+  propagate(node(X, Package), variant_value(Variant, Value2)),
+  Value1 != Value2,
+  node(X, Package) != node(Y, Source),
+
+  % The variant is a singlevalued variant
+  node_has_variant(node(X, Package), Variant, VariantID),
+  not variant_type(VariantID, "multi"),
+
+  % Find the source based on the fact that nothing is propagating to it
+  propagate(node(Y, Source), variant_value(Variant, Value2)),
+  not 1 {propagate(SourceParent, variant_value(Variant, Value2)): depends_on(SourceParent, node(Y, Source))}.
```

Comment:
Curious what @alalazo thinks about this. I wonder if it would be too expensive to pull this out and use it elsewhere, e.g.:

```suggestion
  propagation_source(node(Y, Source), variant_value(Variant, Value2)).
```

and, elsewhere:

```prolog
% a node is the source of propagation if nothing propagated to it
propagation_source(Source, PropagatedValue) :-
  propagate(Source, PropagatedValue),
  not 1 { propagate(Parent, PropagatedValue): depends_on(Parent, Source) }.
```

This is a little clumsy because the `propagate` rules don't preserve any information on what the source was.  We could tweak the encoding in `asp.py` so that the `asp.py` just sets `propagate_source` and the solver derives `propagate` -- them we wouldn't have to be so fancy.

Thoughts @alalazo @becker33?

Reply:
I think my answer would basically be https://github.com/spack/spack/pull/42931#discussion_r1823254801 We should try to get the second argument in `propagate/2` contain info on the source node.

Reply:
So, it seems this might be causing a performance issue after all, see https://github.com/spack/spack/pull/42931#issuecomment-2449224921

Reply:
I'm going to push the diff below, since it passes tests locally, and it performs better when grounding and solving. As an example, with the patch:
```console
$ spack solve --timers trilinos++aggressive_vectorization++shared++debug++mpi
    setup         14.512s
    load           2.219s
    ground         7.778s
    solve          6.861s
    total         32.089s

[ ... ]
```
without the patch:
```console
$ spack solve --timers trilinos++aggressive_vectorization++shared++debug++mpi
    setup         14.511s
    load           2.247s
    ground         9.031s
    solve          7.600s
    total         34.283s

 [ ... ]
```